libyul/Utilities.cpp

```diff
@@ -86,56 +86,96 @@ std::string solidity::yul::reindent(std::string const& _code)
 	return out.str();
 }
 
-u256 solidity::yul::valueOfNumberLiteral(Literal const& _literal)
+LiteralValue solidity::yul::valueOfNumberLiteral(std::string_view const _literal)
 {
-	yulAssert(_literal.kind == LiteralKind::Number, "Expected number literal!");
-
-	static std::map<YulString, u256> numberCache;
-	static YulStringRepository::ResetCallback callback{[&] { numberCache.clear(); }};
+	return LiteralValue{LiteralValue::Data(_literal), std::string(_literal)};
+}
+LiteralValue solidity::yul::valueOfStringLiteral(std::string_view const _literal)
+{
+	std::string const s(_literal);
+	return LiteralValue{u256(h256(s, h256::FromBinary, h256::AlignLeft)), s};
+}
+LiteralValue solidity::yul::valueOfBuiltinStringLiteralArgument(std::string_view _literal)
+{
+	return LiteralValue{std::string(_literal)};
+}
+LiteralValue solidity::yul::valueOfBoolLiteral(std::string_view const _literal)
+{
+	if (_literal == "true")
+		return LiteralValue{1, "true"};
+	else if (_literal == "false")
+		return LiteralValue{0, "false"};
 
-	auto&& [it, isNew] = numberCache.try_emplace(_literal.value, 0);
-	if (isNew)
+	yulAssert(false, "Unexpected bool literal value!");
+}
+LiteralValue solidity::yul::valueOfLiteral(std::string_view const _literal, LiteralKind const& _kind, bool const _unlimitedLiteralArgument)
+{
+	switch (_kind)
 	{
-		std::string const& literalString = _literal.value.str();
-		yulAssert(isValidDecimal(literalString) || isValidHex(literalString), "Invalid number literal!");
-		it->second = u256(literalString);
+	case LiteralKind::Number:
+		return valueOfNumberLiteral(_literal);
+	case LiteralKind::Boolean:
+		return valueOfBoolLiteral(_literal);
+	case LiteralKind::String:
+		return _unlimitedLiteralArgument ? valueOfBuiltinStringLiteralArgument(_literal) : valueOfStringLiteral(_literal);
 	}
-	return it->second;
+	yulAssert(false);
 }
 
-u256 solidity::yul::valueOfStringLiteral(Literal const& _literal)
+std::string solidity::yul::formatLiteral(solidity::yul::Literal const& _literal)
 {
-	yulAssert(_literal.kind == LiteralKind::String, "Expected string literal!");
-	yulAssert(_literal.value.str().size() <= 32, "Literal string too long!");
+	std::string result;
+	if (_literal.value.unlimited())
+		result = _literal.value.builtinStringLiteralValue();
+	else if (_literal.value.hint() && _literal.value.value() == valueOfLiteral(*_literal.value.hint(), _literal.kind).value())
+		result = *_literal.value.hint();
+	else
+		if (_literal.kind == LiteralKind::Boolean)
+		{
+			yulAssert(_literal.value.value() == 0 || _literal.value.value() == 1, "Could not format boolean literal");
```

Comment:
I was indeed using them on invalid literals so that the `AsmAnalysis` would produce proper errors for them instead of failing asserts: Having an invalid literal and then trying to format it into a string which then asserts validity is bound to go wrong. 
But yeah, I have added a `validLiteral` method as you suggested and it does make the code cleaner :) for the `AsmAnalysis` reason I made it an optional (but default: on) assert in `formatLiteral`.